File: mcp-server/test/stub-tasks-api.ts
/**
 * A stand-in for the task API, on a real HTTP server.
 *
 * Deliberately a real endpoint rather than a mocked `TasksClient`: the thing worth testing is the
 * whole path — MCP request, tool dispatch, HTTP call, response translation. A mock at the client
 * boundary would skip the half most likely to break.
 *
 * It reproduces the task API's contract only as far as these tests need: 404 for an unknown id,
 * 400 for a blank title, and PUT as a full replace.
 */

import { createServer, type Server } from 'node:http';
import { randomUUID } from 'node:crypto';

import type { TaskResponse } from '../src/tasks-client.js';

const FIXED_CREATED_AT = '2026-08-23T00:00:00.000000Z';
const FIXED_UPDATED_AT = '2026-08-23T00:00:01.000000Z';

export interface StubTasksApi {
    readonly url: string;
    readonly store: TaskResponse[];
    /** Every request this stub has received. Lets a test prove a code path did NOT call the api. */
    readonly requests: { method: string; path: string }[];
    close(): Promise<void>;
}

export async function startStubTasksApi(): Promise<StubTasksApi> {
    const store: TaskResponse[] = [];
    const requests: { method: string; path: string }[] = [];

    const server: Server = createServer((request, response) => {
        const method = request.method ?? 'GET';
        const url = request.url ?? '/';
        requests.push({ method, path: new URL(url, 'http://stub').pathname });
        void handle(method, url, readBody(request), store)
            .then(({ status, body }) => {
                if (body === undefined) {
                    response.writeHead(status);
                    response.end();
                    return;
                }
                response.writeHead(status, { 'content-type': 'application/json' });
                response.end(JSON.stringify(body));
            });
    });

    await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
    const address = server.address();
    if (address === null || typeof address === 'string') {
        throw new Error('stub task api did not bind a TCP port');
    }

    return {
        url: `http://127.0.0.1:${address.port}`,
        store,
        requests,
        close: () => new Promise<void>((resolve) => server.close(() => resolve())),
    };
}

async function handle(
    method: string,
    url: string,
    body: Promise<unknown>,
    store: TaskResponse[],
): Promise<{ status: number; body?: unknown }> {
    const path = new URL(url, 'http://stub').pathname;
    const id = path.startsWith('/tasks/') ? decodeURIComponent(path.slice('/tasks/'.length)) : undefined;

    if (method === 'GET' && path === '/tasks') {
        return { status: 200, body: store };
    }

    if (method === 'GET' && id !== undefined) {
        const found = store.find((task) => task.id === id);
        return found === undefined ? { status: 404 } : { status: 200, body: found };
    }

    if (method === 'POST' && path === '/tasks') {
        const request = (await body) as Record<string, string | undefined>;
        if (request['title'] === undefined || request['title'].trim() === '') {
            return { status: 400, body: { message: 'title must not be blank' } };
        }
        const created = materialize(randomUUID(), request, FIXED_CREATED_AT, FIXED_CREATED_AT);
        store.push(created);
        return { status: 201, body: created };
    }

    if (method === 'PUT' && id !== undefined) {
        const index = store.findIndex((task) => task.id === id);
        if (index === -1) {
            return { status: 404 };
        }
        const request = (await body) as Record<string, string | undefined>;
        if (request['title'] === undefined || request['title'].trim() === '') {
            return { status: 400, body: { message: 'title must not be blank' } };
        }
        // Full replace, mirroring the real API: omitted fields go back to defaults.
        const existing = store[index]!;
        const replaced = materialize(id, request, existing.createdAt, FIXED_UPDATED_AT);
        store[index] = replaced;
        return { status: 200, body: replaced };
    }

    if (method === 'DELETE' && id !== undefined) {
        const index = store.findIndex((task) => task.id === id);
        if (index !== -1) {
            store.splice(index, 1);
        }
        // 204 whether or not it existed — the api's idempotent delete.
        return { status: 204 };
    }

    return { status: 404 };
}

function materialize(
    id: string,
    request: Record<string, string | undefined>,
    createdAt: string,
    updatedAt: string,
): TaskResponse {
    return {
        id,
        title: request['title']!,
        description: request['description'] ?? null,
        status: request['status'] ?? 'TODO',
        priority: request['priority'] ?? 'MEDIUM',
        createdAt,
        updatedAt,
    };
}

async function readBody(request: NodeJS.ReadableStream): Promise<unknown> {
    const chunks: Buffer[] = [];
    for await (const chunk of request) {
        chunks.push(Buffer.from(chunk));
    }
    const raw = Buffer.concat(chunks).toString('utf8');
    return raw === '' ? {} : JSON.parse(raw);
}

File: mcp-server/src/test/java/dev/petrov/tasks/mcp/StubTasksApi.java
package dev.petrov.tasks.mcp;

import io.micronaut.context.annotation.Requires;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Delete;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Post;
import io.micronaut.http.annotation.Put;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

/**
 * A stand-in for the task API, served from the same context as the MCP server under test.
 *
 * <p>Deliberately a real HTTP endpoint rather than a mocked {@link TasksClient}: the thing worth
 * testing is the whole path — MCP request, tool dispatch, HTTP call, response translation. A mock
 * at the client boundary would skip the half most likely to break.
 *
 * <p>It reproduces the task API's contract only as far as these tests need: 404 for an unknown id
 * and 400 for a blank title.
 */
@Requires(property = "stub.tasks.enabled", value = "true")
@Controller("/tasks")
public class StubTasksApi {

    private final List<TaskDtos.TaskResponse> store = new ArrayList<>();

    @Get
    public List<TaskDtos.TaskResponse> list() {
        return List.copyOf(store);
    }

    @Get("/{id}")
    public HttpResponse<TaskDtos.TaskResponse> get(String id) {
        return store.stream()
                .filter(t -> t.id().equals(id))
                .findFirst()
                .map(HttpResponse::ok)
                .orElseGet(HttpResponse::notFound);
    }

    @Post
    public HttpResponse<TaskDtos.TaskResponse> create(@Body TaskDtos.TaskRequest request) {
        if (request.title() == null || request.title().isBlank()) {
            return HttpResponse.status(HttpStatus.BAD_REQUEST);
        }
        TaskDtos.TaskResponse created = new TaskDtos.TaskResponse(
                UUID.randomUUID().toString(),
                request.title(),
                request.description(),
                Objects.requireNonNullElse(request.status(), "TODO"),
                Objects.requireNonNullElse(request.priority(), "MEDIUM"),
                "2026-08-23T00:00:00.000000Z",
                "2026-08-23T00:00:00.000000Z");
        store.add(created);
        return HttpResponse.created(created);
    }

    @Put("/{id}")
    public HttpResponse<TaskDtos.TaskResponse> update(String id, @Body TaskDtos.TaskRequest request) {
        for (int i = 0; i < store.size(); i++) {
            if (store.get(i).id().equals(id)) {
                // Full replace, mirroring the real API: omitted fields go back to defaults.
                TaskDtos.TaskResponse replaced = new TaskDtos.TaskResponse(
                        id,
                        request.title(),
                        request.description(),
                        Objects.requireNonNullElse(request.status(), "TODO"),
                        Objects.requireNonNullElse(request.priority(), "MEDIUM"),
                        store.get(i).createdAt(),
                        "2026-08-23T00:00:01.000000Z");
                store.set(i, replaced);
                return HttpResponse.ok(replaced);
            }
        }
        return HttpResponse.notFound();
    }

    @Delete("/{id}")
    public HttpResponse<Void> delete(String id) {
        store.removeIf(t -> t.id().equals(id));
        return HttpResponse.noContent();
    }
}
